### PR TITLE
fix(check): CJS types importing dual ESM/CJS package should prefer CJS types

### LINKED ISSUE #16370: Fix referrer node module kind when resolving types file with unspecified extension
See https://github.com/denoland/deno/pull/16332#discussion_r1000494231 -- maybe it's a simple fix of resolving the referrer kind if necessary right there.

### PR DESCRIPTION
**Pull request by dsherret**
Closes #16370

### CODE REVIEW

File: cli/resolver.rs [hunk 1]
```diff
@@ -66,8 +67,7 @@ pub struct ModuleCodeStringSource {
 
 #[derive(Debug)]
 pub struct CliNodeResolver {
-  // not used in the LSP
-  cjs_resolutions: Option<Arc<CjsResolutionStore>>,
+  cjs_resolutions: Arc<CjsResolutionStore>,
```
Comment (dsherret) [inline cli/resolver.rs:70]: This means we have this cache in two places for the LSP, which is not ideal.

Comment (bartlomieju) [reply cli/resolver.rs:70]: I guess we can address this in a follow up?

File: cli/lsp/resolver.rs [hunk 2]
```diff
@@ -477,14 +495,21 @@ async fn create_npm_resolver(
 fn create_node_resolver(
   npm_resolver: Option<&Arc<dyn CliNpmResolver>>,
 ) -> Option<Arc<CliNodeResolver>> {
+  use once_cell::sync::Lazy;
+
+  // it's not ideal to share this across all scopes and to
+  // never clear it, but it's fine for the time being
```
Comment (bartlomieju) [inline cli/lsp/resolver.rs:501]: Add a TODO and open an issue so we don't forget?

**Review by bartlomieju — APPROVED** [review]
LGTM

File: cli/lsp/resolver.rs (hunk 2 above)
Comment (dsherret) [reply cli/lsp/resolver.rs:501]: I initially had one, but removed it because we can fix this once we get a bug that causes us to need to fix it.

**Review by nayeemrmn — APPROVED** [review]
LGTM

**Pull request merged**